src/main/java/com/github/mizool/jediscache/JedisCachingProvider.java

```diff
@@ -0,0 +1,211 @@
+import javax.cache.spi.CachingProvider;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.kohsuke.MetaInfServices;
+
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+
+@Slf4j
+@MetaInfServices(CachingProvider.class)
+public class JedisCachingProvider implements CachingProvider
+{
+    private final JedisPool jedisPool;
+    private WeakHashMap<ClassLoader, HashMap<URI, CacheManager>> cacheManagersByClassLoader;
+
+    public JedisCachingProvider()
+    {
+        String host = System.getProperty("jediscache.host", "127.0.0.1");
+        log.info("JedisCache active, connecting to {}", host);
+        this.jedisPool = new JedisPool(new JedisPoolConfig(), host);
+        this.cacheManagersByClassLoader = new WeakHashMap<>();
+    }
+
+    @Override
+    public synchronized CacheManager getCacheManager(URI uri, ClassLoader classLoader, Properties properties)
+    {
+        if (uri == null)
+        {
+            uri = getDefaultURI();
+        }
+        if (classLoader == null)
+        {
+            classLoader = getDefaultClassLoader();
+        }
+        if (properties == null)
+        {
+            properties = new Properties();
+        }
+
+        HashMap<URI, CacheManager> cacheManagersByURI = cacheManagersByClassLoader.get(classLoader);
+
+        if (cacheManagersByURI == null)
+        {
+            cacheManagersByURI = new HashMap<>();
+        }
+
+        CacheManager cacheManager = cacheManagersByURI.get(uri);
+
+        if (cacheManager == null)
+        {
+            cacheManager = new JedisCacheManager(this, uri, classLoader, properties, jedisPool);
+            cacheManagersByURI.put(uri, cacheManager);
+        }
+
+        if (!cacheManagersByClassLoader.containsKey(classLoader))
+        {
+            cacheManagersByClassLoader.put(classLoader, cacheManagersByURI);
+        }
+
+        return cacheManager;
+    }
+
+    @Override
+    public CacheManager getCacheManager(URI uri, ClassLoader classLoader)
+    {
+        return getCacheManager(uri, classLoader, getDefaultProperties());
+    }
+
+    @Override
+    public CacheManager getCacheManager()
+    {
+        return getCacheManager(getDefaultURI(), getDefaultClassLoader(), null);
+    }
+
+    @Override
+    public ClassLoader getDefaultClassLoader()
+    {
+        return getClass().getClassLoader();
+    }
+
+    @Override
+    public URI getDefaultURI()
+    {
+        try
+        {
+            return new URI(this.getClass().getName());
+        }
+        catch (URISyntaxException e)
+        {
+            throw new CacheException("Failed to create the default URI for jedis cache", e);
+        }
+    }
+
+    @Override
+    public Properties getDefaultProperties()
+    {
+        return null;
+    }
+
+    @Override
+    public synchronized void close()
+    {
+        WeakHashMap<ClassLoader, HashMap<URI, CacheManager>> managersByClassLoader = this.cacheManagersByClassLoader;
```

Comment:
Change variable type to `Map<ClassLoader, Map<URI, CacheManager>>`.

Reply:
Either rename `managersByClassLoader` to `cacheManagersByClassLoader` (which I would prefer) or remove `this.` in this line and the assignment below.